src/hardware/raspberrypi-pico/protocol.c

```diff
@@ -17,25 +17,696 @@
+        num_read+= serial_read_blocking(serial, &(resp[num_read]), cnt-num_read, 10);
+        //sr_spew("rwrsp2 nr %d",num_read);
+
+        if ((num_read < 1)||(num_read>30)) {
+                sr_err("ERROR:Serial_w_resp failed (%d).", num_read);
+                return -1;
+        }else{
+         return num_read;
+        }
+}       
+//Issue a command that expects a single char ack 
+SR_PRIV int send_serial_w_ack(struct sr_serial_dev_inst *serial, char *str){
+        char buf[2];
+        int num_read;
+        //In case we have left over transfer from the device, drain them
+        while((num_read=serial_read_blocking(serial, buf, 2, 10))){
+          //sr_dbg("swack drops 2 previous bytes %d %d",buf[0],buf[1]);
+        }
+        send_serial_str(serial,str);
+        //1000ms timeout
+        num_read =serial_read_blocking(serial, buf, 1, 1000);
+        if ((num_read == 1)&&(buf[0]=='*')) {
+                return SR_OK;
+        }else{
+                sr_err("ERROR:Serial_w_ack %s failed (%d).", str,num_read);
+                if(num_read){
+                  sr_err("ack resp char %c d %d\n\r",buf[0],buf[0]);
+                }
+                return SR_ERR;
+        }
+}        
+
+//Process incoming data stream assuming it is optimized packing of 4 channels or less
+//Each byte is 4 channels of data and a 3 bit rle value, or a larger rle value, or a control signal.
+//This also checks for aborts and ends.
+//If an end is seen we stop processing but do not check the byte_cnt
+//The output is a set of samples fed to process group to perform sw triggering and sending of data to the session
+//as well as maintenance of the serial rx byte cnt.
+//Since we can get huge rle values we chop them up for processing into smaller groups
+//In this mode we can always consume all bytes because there are no cases where the processing of one 
+//byte requires the one after it.
+void process_D4(struct sr_dev_inst *sdi,struct dev_context *d){
+   int32_t j;
+   uint8_t cbyte;
+   uint8_t cval;
+   uint32_t rlecnt=0;
+   uint32_t sampcnt=0; //number of samples received with no rles
+   while(d->ser_rdptr < d->bytes_avail){
+     cbyte=d->buffer[(d->ser_rdptr)];
+     //RLE only byte
+     if(cbyte>=48 && cbyte<=127){ 
+       rlecnt+=(cbyte-47)*8;
+       d->byte_cnt++;
+     }else if(cbyte>=0x80){ //sample with possible rle
+       rlecnt+=(cbyte&0x70)>>4; 
+       if(rlecnt){
+         //On a value change, duplicate the previous values first.
+         //The maximum value of one rle is 640.
+         //To ensure we don't overflow the sample buffer but still send it large chunks of data 
+         //(to make the packet sends to the session efficient) only call process group after
+         //a large number of samples have been seen.
+         //Likely we could use the max rle value of 640 but 2048 gives some extra room.
+         if((rlecnt+d->cbuf_wrptr)>(d->sample_buf_size-2048)){
```

Comment:
Please avoid hardcode 2048, create a macro (#define) to store it.

Reply:
Alan,
Thanks for the review.  Do you have moderator permissions and is this review part of pulling this turnin into the main repo?  I ask because there are a few pico related turnins and it's not clear which if any are being pulled. (I of course consider mine to be the most featured;)).  
Thanks, 
Shawn

Reply:
Hi Shawn, I don't have merge commit bit, but I want to help to fix issues and get things in place. I tried to contact Uwe over Twitter some time ago, but not reply. Maybe someone near him could tell us what happened. It is sad because this project is very good. Maybe @gsigh could help us, he has commit bit.

Reply:
![image](https://user-images.githubusercontent.com/105331211/181909856-9f0fcb46-16e7-48d8-8933-8d2af09730a7.png)

This needs to be in master!